### PR TITLE
Use Slice operation directly on CPU and update RoPE patterns

### PR DESCRIPTION
**Pull request by itikhono**
### Details:
 - SliceToStridedSlice was moved from MOC to Common transformations.
 - Updated RoPE fusion patterns
 - 2 new transformations:  EliminateSlice, SliceSequenceToSingleSlice
 - Updated EliminateStridedSlice transformation to support int32_max case, deleted transformation duplicate from CPU
 - Added new unit tests

Tested locally with these models:

- [x] Means that Model compilation passed, and all RoPE related transformations were applied in the same order
- [x] hf-internal-testing/tiny-random-StableLmForCausalLM
- [x] hf-internal-testing/tiny-random-FalconForCausalLM
- [x] hf-internal-testing/tiny-random-Starcoder2ForCausalLM
- [x] hf-internal-testing/tiny-random-LlamaForCausalLM
- [x] hf-internal-testing/tiny-random-GPTNeoXForCausalLM
- [x] hf-internal-testing/tiny-random-GPTJForCausalLM
- [x] hf-internal-testing/tiny-random-CodeGenForCausalLM
- [x] hf-internal-testing/tiny-random-MistralForCausalLM
- [x] hf-internal-testing/tiny-random-PhiForCausalLM
- [x] Qwen/Qwen1.5-7B
- [x] THUDM/chatglm3-6b
- [x] EleutherAI/gpt-neox-20b
- [x] google/gemma-2b-it
- [x] EleutherAI/gpt-j-6b
- [x] meta-llama/Meta-Llama-3-8B
- [x] mistralai/Mistral-7B-v0.1

### Tickets:
 - *CVS-126971*

### CODE REVIEW

File: src/common/transformations/src/transformations/common_optimizations/optimize_strided_slice.cpp [hunk 1]
```diff
@@ -366,25 +368,72 @@ bool ov::pass::GroupedSliceToVSplitOptimization::run_on_model(const std::shared_
...
+    auto const_axes_2_pattern = wrap_type<v0::Constant>();
+    auto slice_1_pattern =
+        wrap_type<v8::Slice>({any_input(), any_input(), any_input(), any_input(), const_axes_1_pattern},
+                             consumers_count(1));
+    auto slice_2_pattern =
+        wrap_type<v8::Slice>({slice_1_pattern, any_input(), any_input(), any_input(), const_axes_2_pattern});
+
+    ov::matcher_pass_callback callback = [=](Matcher& m) {
+        const auto& pattern_to_output = m.get_pattern_map();
+        auto slice_1 = pattern_to_output.at(slice_1_pattern);
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/optimize_strided_slice.cpp:387]: You're using get_pattern_map() here. I saw, other transformations make use of get_pattern_value_map(). As I understand, there's no big difference as the first maps to nodes, the latter to output. Or are there any other reasons to use one over another?

**Review by CuriousPanCake — APPROVED** [review]

File: src/common/transformations/include/transformations/common_optimizations/optimize_strided_slice.hpp [hunk 2]
```diff
@@ -56,6 +57,15 @@ class ov::pass::GroupedSliceToVSplitOptimization : public ov::pass::ModelPass {
     bool run_on_model(const std::shared_ptr<ov::Model>& m) override;
 };
 
+/**
+ * TBA
```
Comment (itikhono) [inline src/common/transformations/include/transformations/common_optimizations/optimize_strided_slice.hpp:61]: please add a description

File: src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp [hunk 3]
```diff
@@ -837,9 +837,38 @@ ov::pass::NopSliceBeforeGatherElements::NopSliceBeforeGatherElements() {
...
+    auto begin_const = pattern::wrap_type<ov::op::v0::Constant>();
+    auto end_const = pattern::wrap_type<ov::op::v0::Constant>();
+    auto step_const = pattern::wrap_type<ov::op::v0::Constant>();
+    auto axes = pattern::any_input();
+    auto pattern = pattern::wrap_type<ov::op::v8::Slice>({input, begin_const, end_const, step_const, axes});
+
+    ov::matcher_pass_callback matcher_pass_callback = [=](pattern::Matcher& m) {
+        auto slice = std::dynamic_pointer_cast<ov::op::v8::Slice>(m.get_match_root());
+
+        int64_t max_int = slice->input_value(2).get_element_type() == element::i32
```
Comment (evkotov) [inline src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp:853]: only element::i32? what about another i[0-9]+ and u[0-9]+? i64, u32, ...

File: src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp [hunk 4]
```diff
@@ -903,14 +932,17 @@ ov::pass::NopStridedSlice::NopStridedSlice() {
 
         auto end_node = strided_slice_node->get_input_node_shared_ptr(2);
         if (const auto& end_constant_node = ov::util::get_constant_from_source(end_node)) {
+            int64_t max_value = end_node->get_element_type() == ov::element::i32 ? std::numeric_limits<int32_t>::max()
```
Comment (evkotov) [inline src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp:935]: Could it be another interger type? i64, u32....?

File: src/common/transformations/src/transformations/common_optimizations/optimize_strided_slice.cpp [hunk 5]
```diff
@@ -366,25 +368,72 @@ bool ov::pass::GroupedSliceToVSplitOptimization::run_on_model(const std::shared_
...
+                return false;
+            }
+        }
+
+        auto begin = std::make_shared<v0::Concat>(OutputVector{slice_1->input_value(1), slice_2->input_value(1)}, 0);
+        auto end = std::make_shared<v0::Concat>(OutputVector{slice_1->input_value(2), slice_2->input_value(2)}, 0);
+        auto step = std::make_shared<v0::Concat>(OutputVector{slice_1->input_value(3), slice_2->input_value(3)}, 0);
+        auto axes = std::make_shared<v0::Concat>(OutputVector{slice_1->input_value(4), slice_2->input_value(4)}, 0);
+        auto one_slice = std::make_shared<ov::op::v8::Slice>(slice_1->input_value(0),
+                                                             try_fold_unary_output(begin),
```
Comment (evkotov) [inline src/common/transformations/src/transformations/common_optimizations/optimize_strided_slice.cpp:408]: is folding needed here? we have ConstantFolding transformation in MOC

File: src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp (hunk 4 above)
Comment (evkotov) [reply src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp:935]: Could we have here element_type::u64 and std::numeric_limits<unt64_t>::max()
std::numeric_limits<int64_t>::max() != std::numeric_limits<uint64_t>::max()

File: src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp (hunk 3 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp:853]: no, both i32 and i64 are supported.

File: src/common/transformations/src/transformations/common_optimizations/optimize_strided_slice.cpp (hunk 5 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/optimize_strided_slice.cpp:408]: it's required

File: src/common/transformations/src/transformations/common_optimizations/optimize_strided_slice.cpp (hunk 1 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/optimize_strided_slice.cpp:387]: discussed offline

File: src/common/transformations/src/transformations/common_optimizations/common_optimizations.cpp [hunk 6]
```diff
@@ -125,7 +126,9 @@ bool ov::pass::CommonOptimizations::run_on_model(const std::shared_ptr<ov::Model
 
     using namespace ov::pass;
     REGISTER_PASS(manager, DisableDecompressionConvertConstantFolding)
-
+    // MOCTransformations contain StridedSliceOptimization transformation,
+    // so we must call SliceToStridedSlice before MOCTransformations call
+    REGISTER_PASS(manager, SliceToStridedSlice, true)
```
Comment (jane-intel) [inline src/common/transformations/src/transformations/common_optimizations/common_optimizations.cpp:131]: Q: How would we lower this pass down the pipeline in the future?

File: src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp [hunk 7]
```diff
@@ -837,9 +837,38 @@ ov::pass::NopSliceBeforeGatherElements::NopSliceBeforeGatherElements() {
...
+ov::pass::EliminateSlice::EliminateSlice() {
+    MATCHER_SCOPE(EliminateSlice);
 
-ov::pass::NopStridedSlice::NopStridedSlice() {
-    MATCHER_SCOPE(NopStridedSlice);
+    auto input = pattern::any_input();
+    auto begin_const = pattern::wrap_type<ov::op::v0::Constant>();
+    auto end_const = pattern::wrap_type<ov::op::v0::Constant>();
+    auto step_const = pattern::wrap_type<ov::op::v0::Constant>();
+    auto axes = pattern::any_input();
```
Comment (jane-intel) [inline src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp:847]: My understanding is that axes is an optional input and could be missing meaning all the axes are involved in slicing.

File: src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp [hunk 8]
```diff
@@ -923,47 +955,52 @@ ov::pass::NopStridedSlice::NopStridedSlice() {
...
+    auto end = pattern::any_input();
     auto optional_stride_const = pattern::wrap_type<ov::op::v0::Constant>();
-    auto pattern = pattern::wrap_type<ov::op::v1::StridedSlice>({input, begin_const, end_const, optional_stride_const});
+    auto strided_slice = pattern::wrap_type<ov::op::v1::StridedSlice>({input, begin, end, optional_stride_const});
+
+    auto axes = pattern::any_input();
+    auto slice = pattern::wrap_type<ov::op::v8::Slice>({input, begin, end, optional_stride_const, axes});
+    auto pattern = std::make_shared<pattern::op::Or>(OutputVector{strided_slice, slice});
 
     ov::matcher_pass_callback matcher_pass_callback = [=](pattern::Matcher& m) {
```
Comment (jane-intel) [inline src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp:971]: Overall, I agree with the idea of this and other Elimination passes here. However, they seem complex, so I strongly recommend rewriting them using Symbolic infra in the future.

File: src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp (hunk 4 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp:935]: yes, you are right
I will improve it in the next PR

**Review by jane-intel — COMMENTED** [review]
Nice job

**Review by jane-intel — APPROVED** [review]

File: src/common/transformations/include/transformations/common_optimizations/optimize_strided_slice.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/include/transformations/common_optimizations/optimize_strided_slice.hpp]: fixed

Comment (itikhono) [reply src/common/transformations/include/transformations/common_optimizations/optimize_strided_slice.hpp]: fixed

**Pull request merged**